tests/editing/totals_and_components/test_totals_and_components.py

```diff
@@ -6378,17 +6527,64 @@ def test_correct_components(
             )
 
 
-def compare_list_of_floats(components: List[float], expected_components: List[float]):
+def compare_results_to_expected_results(
+    results: TotalsAndComponentsOutput, expected_result: tuple
+):
+    assert results.identifier == expected_result[0]
+
+    if results.absolute_difference is None:
+        assert results.absolute_difference == expected_result[1]
+    else:
+        assert round(results.absolute_difference, 7) == round(expected_result[1], 7)
+
+    if results.low_percent_threshold is None:
+        assert results.low_percent_threshold == expected_result[2]
+    else:
+        assert round(results.low_percent_threshold, 7) == round(expected_result[2], 7)
+
+    if results.high_percent_threshold is None:
+        assert results.high_percent_threshold == expected_result[3]
+    else:
+        assert round(results.high_percent_threshold, 7) == round(expected_result[3], 7)
+
+    if results.final_total is None:
+        assert results.final_total == expected_result[4]
+    else:
+        assert round(results.final_total, 7) == round(expected_result[4], 7)
+
+    assert results.tcc_marker == expected_result[6]
+
+    if results.tcc_marker == "T" or results.tcc_marker == "C":
+        sum_of_components = 0
+        for component in results.final_components:
+            print(f"fc - {results.final_components}")
+            if not math.isnan(component):
+                sum_of_components += Decimal(str(component))
+
+        assert round(sum_of_components, 7) == round(expected_result[4], 7)
+
+    compare_list_of_decimals(results.final_components, expected_result[5])
+
+
+def compare_list_of_decimals(
+    components: List[Decimal], expected_components: List[Decimal]
+):
     for i, (component, expected) in enumerate(zip(components, expected_components)):
         if math.isnan(component) and math.isnan(expected):
-            assert math.isnan(component) and math.isnan(expected), f"Both components at index {i} are NaN"
+            assert math.isnan(component) and math.isnan(
+                expected
+            ), f"Both components at index {i} are NaN"
         elif math.isnan(component) or math.isnan(expected):
-            assert False, f"Component at index {i} is NaN, the other is not component {component}, expected {expected}"
-        elif component != expected:
-            assert False, f"Values of components at index {i} do not match: component {component}, expected {expected}"
+            assert (
+                False
+            ), f"Component at index {i} is NaN, the other is not component {component}, expected {expected}"
+        elif round(component, 7) != round(expected, 7):
```

Comment:
What are we doing here? The rounding isn't always 7 digits is it?

Reply:
No but it is most commonly 6 or 7 digits for Ellie's examples. Happy to change it to another number if you like :)

Reply:
It is taking the values of the components to 7.d.p and checking they match

Reply:
Otherwise we get assertion errors e.g 32.9999999999999999999999999999999999999999999, expected 33

Reply:
I think we should have a call to work through this. I want to make sure the numbers we are getting out using Decimal are appropriate before any rounding takes place. When we talked about adding rounding to the tests I assumed you would look at the sheet and dynamically apply the rounding that Ellie was using - e.g sheet b6 row 1 final component has rounding to 4 places, row 2 has rounding to 3 places and row 3 has rounding to 5 places.